magma/spec/query_spec.rb

```diff
@@ -107,15 +144,10 @@ def update(revisions, user_type=:editor)
     end
 
     it 'supports ::count' do
-      hydra = create(:labor, :hydra)
-      stables = create(:labor, :stables)
-      lion = create(:labor, :lion)
-      hind = create(:labor, :hind)
-
-      poison = create(:prize, labor: hydra, name: 'poison', worth: 5)
-      poop = create(:prize, labor: stables, name: 'poop', worth: 0)
-      iou = create(:prize, labor: stables, name: 'iou', worth: 2)
-      skin = create(:prize, labor: lion, name: 'skin', worth: 6)
+      poison = create(:prize, labor: @hydra, name: 'poison', worth: 5)
+      poop = create(:prize, labor: @stables, name: 'poop', worth: 0)
+      iou = create(:prize, labor: @stables, name: 'iou', worth: 2)
+      skin = create(:prize, labor: @lion, name: 'skin', worth: 6)
 
       query(['labor', '::all', 'prize', '::count' ])
 
```

Comment:
During development, this was the only test that failed for actual reasons, not simply because the test data was now disconnected. At first I added the disconnected data filtering to ModelPredicate, the same way as the `restricted` filtering operates. However, this test began to fail if disconnected data was filtered, because it would no longer return the entry `[ 'Ceryneian Hind', 0 ]`. It turned out that this was because the filtering added two non-disconnected requirements to the same table, by transforming the original query to look like this:
`[ 'labor', [ '::has', 'project' ], '::all', 'prize', [ '::has', 'labor' ], [ 'labor', '::has', 'project' ], '::count' ]`

Somehow these two requirements intersect and drop out the row (somehow, I imagine, due to left outer-joining an empty row to itself).

My solution here is to impose the requirement only on the first clause; the presumption is that if that is connected, then anything joined to that must also be okay (that is, not disconnected). This logic might fall apart for link attributes (which could be disconnected but attached via a link to some connected data), though, so orphaned data might sneak into the query results anyway.

I'm more disturbed by the failure of the left outer join, and the removal of empty rows of data, which is a pretty crucial ability for querying. While I am papering over the problem here by creating a better filter, there is nothing preventing someone from making the query above, and it is not obvious why it should result in a count == 0 row being removed. There are some real deep issues lurking behind the relational algebra of this query logic that must eventually be confronted.